src/core/storage/sframe_data/join_impl.cpp

```diff
@@ -156,67 +161,154 @@ hash_join_executor::hash_join_executor(const sframe &left,
     _right_join = true;
   }
 
+  ASSERT_EQ(_left_join_positions.size(), _right_join_positions.size());
+
+  /* handy script for reverse lookup */
+  for(size_t i = 0; i < _left_join_positions.size(); ++i) {
+    auto ret = _right_to_left_join_positions.emplace(_right_join_positions[i],
+                                                     _left_join_positions[i]);
+    ASSERT_TRUE(ret.second);
+
+    ret = _left_to_right_join_positions.emplace(_left_join_positions[i],
+                                                _right_join_positions[i]);
+    ASSERT_TRUE(ret.second);
+  }
+
+  /*
+   * using names from sframe: unique and non-empty
+   * the new names in result sframe should have no confilict
+   *
+   * this is order dependent; doing in reverse will result in
+   * different result.
+   *
+   * since _alter_name_right only applies to original order,
+   * we have to obtain the original construction
+   **/
+
+  // get orginal left sframe's column names
+  auto col_names_original_order = left.column_names();
+  _output_column_names = left.column_names();
+  _output_column_types = left.column_types();
+
+  std::set<std::string> new_table = {
+      std::make_move_iterator(col_names_original_order.begin()),
+      std::make_move_iterator(col_names_original_order.end())};
+
+  // get orginal right sframe's column names
+  col_names_original_order = right.column_names();
+  // check the original right sframe columns
+  for (auto& name : col_names_original_order) {
+    // skip join_keys
+    if (_right_to_left_join_positions.count(right.column_index(name)))
+      continue;
+
+    if (new_table.count(name)) {
+      if (_alter_names_right.count(name)) {
+        /*
+         * user defined resolution shall not have any conflict with
+         * all col names visited so far; but can be the same to col
+         * names haven't seen so far.
+         **/
+        auto itr = _alter_names_right.find(name);
+        if (!new_table.insert(itr->second).second) {
+          std::stringstream ss;
+          ss << "user provided column name { " << itr->second << " } conflicts with table name used in SFrame";
+          log_and_throw(ss.str());
+        }
+        _output_column_types.push_back(right.column_type(right.column_index(name)));
+        _output_column_names.push_back(itr->second);
+        new_table.insert(itr->second);
+      } else {
+        _output_column_types.push_back(right.column_type(right.column_index(name)));
+        // default collision resolv. see SFrame::generate_valid_column_name.
+        // if SFrame::generate_valid_column_name changes, this will break.
```

Comment:
I'm pretty sure the order is applied as the same as what SFrame does. Under the hood, first in and first serves. @TobyRoseman you can actually verify it by reading `SFrame::generate_valid_column_name`.

Reply:
Column order is deterministic.

Reply:
@ylow - I'm talking about column order across sessions. The column order can't be deterministic if a dictionary is used to create the SFrame.

Here is the use case I’m concerned about: suppose a user has a script that they run regularly. The script creates some Sframes. Then does a merge using this new feature. Then does some more stuff and makes assumptions about the names of columns. This script could potentially break because the column names across session is not deterministic.

Reply:
It is true that SFrame order will not be deterministic if I do
```
tc.SFrame({'a':[1,2,3],'b':[2,3,4]...})
```
But since an SFrame cannot contain two columns of the same name, it is unclear to me that this causes issues? Can you come up with an example?

Reply:
@ylow https://github.com/apple/turicreate/issues/2129

the name resolution {A->B, B->C} assumes A appears ahead of B in SFrame. But if you do
```
tc.SFrame({'A':[1,2,3],'B':[2,3,4]...})
```
Then it's non-deterministic that A appears before B.

Reply:
@TobyRoseman @ylow 
SFrame left: key, A
SFrame right: Key, A, B
if we resolve in an order A->B, then B->C, the output of join on `Key` is Key, A, B, C.

but somehow we have the right frame built from a dict to be
SFrame right: Key, B, A.
Then it will raise exception:
> Key, A
> Key, A, B
> Key, A, B, <- adding A, conflict found, replace A to B, conflict found again, raise an exception. 

So the resolution depends on the order in SFrame. One way to resolve this is to make the operation order independent. The other is to make sfame construction deterministic.

Reply:
This feature is no longer used due to making the conflict resolution independent.